### DISCUSSION TITLE
Visual Studio and CMake

### DISCUSSION THREAD

**Discussion by pappde** [pappde/bmai#12] [General]
(follow-up to comment posted in other discussion)

To facilitate debugging, I recommend using Visual Studio, which has an excellent debugger. There is a free version, and I believe there is an online version as well.  

Actually, I tried to load the CMake in Visual Studio and ran into some issues (I have not used CMake). 

(1) CMAKE VERSION
The first issue is that Visual Studio 2017 only runs CMake 3.12 and the CMakeLists.txt requires 3.14. It seems "safe" to change the required CMake version to an earlier version, so I will look to do so.

(2) CMAKESETTINGS.JSON
I did notice that VS wants to change to a CMakeSettings.json. Will this be compatible? 

Alternatively, I could always add a SLN/VCPROJ but it looks like Visual Studio will support debugging and compiling from the IDE using a CMake so for compatibility may be best to stick with that. 

Example JSON:
{
  "configurations": [
    {
      "name": "x64-Debug",
      "generator": "Ninja",
      "configurationType": "Debug",
      "inheritEnvironments": [
        "msvc_x64_x64"
      ],
      "buildRoot": "${env.USERPROFILE}\\CMakeBuilds\\${workspaceHash}\\build\\${name}",
      "installRoot": "${env.USERPROFILE}\\CMakeBuilds\\${workspaceHash}\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "-v",
      "ctestCommandArgs": ""
    }
  ]
}

(3) RELEASE CONFIGURATION
I noticed there is only an x64-Debug build, which may be a Visual Studio default behavior.  It would be good to have Debug and Release configurations.

**Comment by pappde** [comment]
(4) EXE NAME
I noticed it is compiling as "Project.exe". Where does "PROJECT_NAME" and "SOURCE" come from in the CMakeLists.txt?

**Reply by danlangford** [reply]
4) when i run Cmake and make on the command line it comes out as a binary "bmai" (bmai.exe on windows). ill import and compile via visual studio and try to see what is going on here. sorry for the hassle.

**Reply by danlangford** [reply]
update: i think you accidentally got rid of "project(bmai)" from the cmake file. I'm working on a PR ill slip it back in

**Reply by pappde** [reply]
> update: i think you accidentally got rid of "project(bmai)" from the cmake file. I'm working on a PR ill slip it back in

I just saw the commit - thanks for fixing that. Presumably that explains why I saw "project.exe"!

**Comment by danlangford** [comment]
1) ya you are welcome to take cmake versions down to a compatible version
2) cmakesettings.json is specific to visualstudio. its visual studio's adaption layer between what it knows and what cmake knows. you are welcome to let you visual studio edit that file to your hearts content. for the best cross platform experience (I do compile and run BMAI on MacOS and Linux and Windows) you will not want to edit the cmkaesettingsjson directly but try to drive your changes through the CMake file and let the IDE update the json as it deems necessary
3) ill have to read up on the best way to represent debug vs release builds. i think I read that there may not be a good cross platform approach. ill try to follow up on this

**Reply by pappde** [reply]
3. per other discussion about "assert()" (#14), you  definitely want to have the concept of a release/optimized build vs a debug/development build.  The two reasons:
- compile out assert(), which should only be validating logic and not input
- enabling compiler optimizations

BMAI is computationally intensive (ideal for multi-threaded) so I expect you would see significant difference in runtime.
